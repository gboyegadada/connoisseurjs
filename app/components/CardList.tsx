import React, { ReactNode, FC, useState } from 'react'
import Card from './Card'
import { ResultItem } from '../types/search'
import ReactPaginate from 'react-paginate'
import { createPaginatorConnect } from '../redux/store'
import { MdInfoOutline, MdSwapVert } from 'react-icons/md'

interface CardListProps {
    children?: ReactNode
    list: ResultItem[]
}

const PaginatorConnect = createPaginatorConnect({
    mapActions: actions => ({
        next: actions.next,
        previous: actions.previous,
        gotoPage: actions.gotoPage
    }),
    mapState: selectors => ({
        offset: selectors.getOffset(),
        pageLength: selectors.getPageLength(),
        pageCount: selectors.getPageCount(),
        totalCount: selectors.getTotalCount(),
        itemCount: selectors.getItemCount(),
        q: selectors.getSearchQuery(),
        queryParams: selectors.getQueryParams()
    })
})

const CardList: FC<CardListProps> = (props) => {
    const { list } = props

    const [pageCount, setPageCount] = useState(0)
    const numFormat = new Intl.NumberFormat('en-US');

    return (
        <PaginatorConnect>
            {(_, actions) =>(
            <>
                <div className='top'>
                    <ul className='list-info'>
                        <li>
                            <MdInfoOutline size='1.2em' /> Items <strong>{(_.offset+1)}</strong> to <strong>{(_.offset + _.itemCount)}</strong> of <strong>{numFormat.format(_.totalCount)}</strong>
                        </li>

                        <li>
                            <MdSwapVert size='1.2em' />
                        </li>
                    </ul>
                </div>
                
                <ul className='grid card-list'>
                    {list.map(item => (
                        <Card key={item.uniqueId} data={item} />
                    ))}
                </ul>
            
                <div className='bottom flex-center'>
                <ReactPaginate
                    previousLabel={'previous'}
                    nextLabel={'next'}
                    breakLabel={'...'}
                    breakClassName={'break-me'}
                    pageCount={_.pageCount}
                    marginPagesDisplayed={2}
                    pageRangeDisplayed={5}
                    onPageChange={pg => actions.gotoPage({ page: pg.selected, query: {...{q: _.q}, ..._.queryParams} })}
                    containerClassName={'pagination'}
                    activeClassName={'active'}
                />
                </div>
            </>
            )}
        </PaginatorConnect>
    )
}

export default CardList