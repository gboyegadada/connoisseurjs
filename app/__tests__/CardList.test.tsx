import React from 'react';
import configureStore from 'redux-mock-store'
import { mount } from 'enzyme'

import CardList from "../components/CardList";
import { Provider } from 'react-redux';

import state__updateSearchResults from './__mock__/states/updateSearchResults.json'

jest.mock('react-router-dom');

describe("<CardList />", () => {
  const initialState = state__updateSearchResults
  const mockStore = configureStore()
  let store: any 
  let wrapper: any

  beforeEach(()=>{
      store = mockStore(initialState)
      wrapper = mount(<Provider store={store}><CardList 
              list={state__updateSearchResults.response.results}
          /></Provider>)
  })

  test("should display card grid", async () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('PaginationBoxView').length).toEqual(1)
  })
})