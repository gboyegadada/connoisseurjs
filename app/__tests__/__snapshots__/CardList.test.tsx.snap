// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardList /> should display clickable Facet Value 1`] = `
<ConnectAdvanced(reduxRenderPropRender)>
  <Component />
</ConnectAdvanced(reduxRenderPropRender)>
`;
